### DISCUSSION TITLE
Source generator for IDisposable

### DISCUSSION THREAD

**Discussion by HakamFostok** [dotnet/runtime#79801] [Ideas]
There is a lot of boilerplate to implement the IDisposable interface correctly, the [docs](https://learn.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose) describes that well, but still it is a lot of boilerplate.

I think it would be much easier if this is done like the source generator
here is an example

```
[DisposableAttribute]
public partial class MyClass
{
     public partial void CleanManaged()
     { 
            // cleaning managed code
     }

     public partial void CleanUnmanaged()
     { 
            // cleaning unmanaged code
     }
}
````

I have been trying to achieve something like this, and I have some results that could be considered good in [this repo](https://github.com/HakamFostok/Disposer)
but of course, the .net has much higher standards than my code, It is just a proof of concept

**Comment by svick** [comment]
Personally, I'd prefer going kind of in the opposite direction: recommend a simple version of `Dispose()`, which works fine in most cases and only use the more complicated version when necessary. (I've talked about this before at https://github.com/dotnet/docs/issues/5485.)

In particular, the commonly recommended pattern with `Dispose(bool)` is only necessary if you need to support:

1. inheritance
2. finalizers/raw unmanaged resources

Inheritance is still common (though less so than in the early days of .Net), but finalizers and raw unmanaged resources should almost never be used. Which is why I think that a simpler version of the disposable pattern should be recommended, instead of hiding the unnecessarily complex version behind a source generator.
